practical-penguins/trivia_tavern/twilio_messenger/views.py

```diff
@@ -0,0 +1,237 @@
+
+            SMSBot.send(goodbye, player.phone_number)
+            SMSBot.send(player.get_answers(), player.phone_number)
+            player.delete()
+        return tally_results
+
+    @staticmethod
+    def send_quiz_invite(number, active_trivia_quiz):
+        intro = (f'Hello! You\'ve been invited to play {active_trivia_quiz.trivia_quiz.name}'
+                 f'If you\'re not game, just text back !quit we won\'t bother you!'
+                 f'Otherwise, please text back a team name to join')
+        SMSBot.send(intro, number)
+        SMSBot.register(number, active_trivia_quiz)
+
+
+@csrf_exempt
+def sms_reply(request):
+    """sms_reply is a handler method that triggers when the url 'sms' is
+    navigated to. Your Twilio account should be configured to point to:
+    <this-site>/sms so that those texts will be recieved here and processed
+    by this function. You can think of this like a 'main' method
+
+    For testing purposes, the developer recommends ngrok to host a temporary
+    server and domain name that points to your localhost
+    (for example http://a2b0fd3c60b6.ngrok.io) and set your
+    twilio account settings in Phone Numbers > Manage Numbers > Active numbers
+    under "Messaging" and set a webhook to <url>/sms/ BUT do not forget the
+    trailing slash!
+    """
+    # Get details about the message that just came in
+    from_ = request.POST.get('From', None)
+    body = request.POST.get('Body', None)
+
+    # check if the text is from a registered Player, can be null
+    player = Player.objects.filter(phone_number=from_)
+    available_quizzes = ActiveTriviaQuiz.objects.all()
+    session_codes = [q.session_code for q in available_quizzes]
+
+    if body.upper() == '!QUIT':
+        SMSBot.player_quit(player)
+        return redirect('/')
+
+    if player.exists():
+        player = player.first()
+        player_quiz = player.active_quiz
+        if player.team_name == '':
+            team_names = [t.team_name for t in Player.objects.all()]
+            if body in team_names:
+                SMSBot.send('Sorry that name is taken!', from_)
+            else:
+                SMSBot.pick_team(body, player)
+
+        elif player_quiz.current_question_index == 0:
+            SMSBot.pre_quiz(body, player)
+
+        else:
+            # Player is answering the question
+            SMSBot.evaluate_answer(body, player)
+            # Optional, send players their score after every question
+            # SMSBot.send(f'Your current score is: {player.points}/{len(question_set)}', from_)
+
+    elif body in session_codes:
+        fetch_quiz = available_quizzes.get(session_code=body)
+        SMSBot.register_with_code(from_, fetch_quiz)
+
+    else:
+        msg = ('This number has not started any quizzes. '
+               'Please send a valid session code to start!'
+               )
+        SMSBot.send(msg, from_)
+
+    # no page to display, sorry :(, redirect to another
+    return redirect('/')
```

Comment:
overall, nicely written. it does feel a bit verbose to read. I'm not sure its a bad thing and I'm not sure if factoring out some business logic would make it that much better.

Reply:
Yeah, I think it could have been separated into one other function, potentially starting from line 207. But it does read quite well and isn't *too* long, so I don't think it's 100% necessary in this case.